### PR TITLE
Taproot tweaks generalization & KeyPair support

### PR DESCRIPTION
**Pull request by dr-orlovsky**
* Adds taproot-related methods to `Script`
* Fixes API for existing taproot methods
* Generalizes `TapTweak` trait to work with both public keys and key pairs

~~UPD: PR is pending https://github.com/rust-bitcoin/rust-secp256k1/pull/342~~

### CODE REVIEW

File: src/blockdata/script.rs
Comment (tcharding) [inline src/blockdata/script.rs:458]: This line should probably be indented also with the above change.

**Comment by dr-orlovsky** [comment]
Rebased and addressed nits by @tcharding

**Comment by tcharding** [comment]
> Rebased and addressed nits by @tcharding

Thanks

**Comment by dr-orlovsky** [comment]
@sanket1729 @kixunil can you pls review this PR since it is quite important for completion of Taproot support - we need to tweak not only public keys, but also private ones, if we are going to even spent taproot key-path :)

**Comment by Kixunil** [comment]
I wasn't able to do deep, high-focus reviews lately but I will try to find some time for it today/tomorrow.

File: src/util/schnorr.rs
Comment (sanket1729) [reply src/util/schnorr.rs:75]: I think everything would be cleaner if you 
```rust
pub trait TapTweak<T> {
	fn tap_tweak<C : Verification>(&self, secp: &Secp256k1<C>, merkle_root: Option<TapBranchHash>) -> T;
}

impl TapTweak<TweakedPublicKey> for UnTweakedKey {
/	/ current impl as is
}

impl TapTweak<TweakedKeyPair> for KeyPair {
	fn tap_tweak<C : Verification>(&self, secp: &Secp256k1<C>, merkle_root: Option<TapBranchHash>) -> T {
		self.tweak_add_assgn(..params); // this does not require Signing context at all and should make this clean
	}
}
```

**Review by sanket1729 — COMMENTED** [review]
Before proceeding with the above review, we should form consensus on the one comment.

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs:75]: Correct; I was confused by generalization of the return type. Actually this way we can use type alias, which will be more correct. Will re-do

Comment (dr-orlovsky) [reply src/util/schnorr.rs:75]: Ah, I recall why I did this. Originally I chosen the path you proposed, but it ended up with additional `C: Signing` requirement for the context to tweak the `KeyPair`, which started polluting `Address` API after (which actually does not need a signing context) etc.

Comment (sanket1729) [reply src/util/schnorr.rs:75]: See: https://github.com/rust-bitcoin/rust-secp256k1/pull/342/commits . I think this should resolve everything here

Comment (dr-orlovsky) [reply src/util/schnorr.rs:75]: This is what we need. But now we need to wait for the new secp release to proceed with this PR :(

Comment (sanket1729) [reply src/util/schnorr.rs:75]: Can you separate out other parts of the PR. I think we can proceed with those. We can leave the `TapTweak` impl after we have a secp release.

Comment (dr-orlovsky) [reply src/util/schnorr.rs:75]: Done: #721

Removing that part from here

**Comment by Kixunil** [comment]
Sadly today wasn't great and I don't expect I'll be able to. If I get lucky it'll be tomorrow morning. Otherwise not sooner than Tuesday :(

**Comment by dr-orlovsky** [comment]
@Kixunil don't worry; this PR got blocked by the required upstream changes (see my discussion with @sanket1729 above). The rest of stuff has moved to #721 which should be simpler to review.

**Comment by dr-orlovsky** [comment]
https://github.com/rust-bitcoin/rust-bitcoin/pull/721#discussion_r760044820

**Comment by dr-orlovsky** [comment]
Will re-base and use the new secp library once the pending issue with it will be fixed in a new point release (I am talking about #360)

**Comment by dr-orlovsky** [comment]
Fixed, rebased, ready for review

File: src/util/schnorr.rs
Comment (sanket1729) [inline src/util/schnorr.rs:51]: In 8b4c5b9d5d0c11b13a2f0cfbe26d20cd7cf24a86
Is it a good idea to have this? I would rather have users parse `UnTweakedPubKey` and call assume_dangerous.

File: src/util/schnorr.rs
Comment (sanket1729) [inline src/util/schnorr.rs:64]: In d2cdbddd0b57452f6b0564f1ce0da3b39a684a34
nit: The documentation for the trait should be updated here to also state it allows keypair tweaking, The description of the tweaking can also be updated with one line stating that for keypairs the secret keys inside the keypair is also tweaked.

File: src/blockdata/script.rs
Comment (sanket1729) [reply src/blockdata/script.rs]: nit: I think this might have been mistakenly "resolved"

File: src/blockdata/script.rs
Comment (sanket1729) [inline src/blockdata/script.rs:318]: In 234d9a49aec9a272857e5a4148d4cb5a2f1a6186
nit: requesting to add a deprecating change here since it is changing the function name

File: src/blockdata/script.rs
Comment (sanket1729) [inline src/blockdata/script.rs:323]: In 234d9a49aec9a272857e5a4148d4cb5a2f1a6186
nit: Deprecating change would be better because it changes the function name

File: src/blockdata/script.rs
Comment (sanket1729) [reply src/blockdata/script.rs]: This is resolved in the next commit. Resolving it.

**Review by sanket1729 — COMMENTED** [review]
Left some comments. CI is also failing on no_std.

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs:51]: `FromStr` is required by serde impl. Also, this brings us back to the question raised in #737.

**Comment by dr-orlovsky** [comment]
@sanket1729 addressed your requests except the first one, where I gave my explanation why `FromStr` is still required.

File: src/util/schnorr.rs
Comment (Kixunil) [reply src/util/schnorr.rs:51]: I suggest don't impl (De)Serialize until we are not confident what the representation should be. We can always add impl later without breaking anything.

Comment (dr-orlovsky) [reply src/util/schnorr.rs:51]: If I am not wrong `TweakedPublicKey` should be used in PSBT, and removing serde there will be a regression

Comment (apoelstra) [reply src/util/schnorr.rs:51]: I think it's okay to parse as `TweakedPublicKey`. I agree that there's a safety loophole here because we can't actually distinguish tweaked keys from untweaked keys, but I think it would be too unergonomic (and as @dr-orlovsky says, possibly incompatible with the PSBT API) to make the user do the `assume_dangerous` dance every time they parse a key.

Especially because untweeaked keys will not exactly be common in the wild .. so we are asking the user to think about an undetectable rare situation that they may not even understand.

File: src/util/schnorr.rs
Comment (apoelstra) [inline src/util/schnorr.rs:73]: In 73f7e96ab15308afc4c1674068044edc38bd60db:

There are trailing spaces here and on two other lines in this file.

File: src/blockdata/script.rs
Comment (apoelstra) [inline src/blockdata/script.rs:341]: In e49a6d98f5b1bdf7bafe130777f9d8ce094d4760:

I think this name should be reserved for a conversion from a `TweakedKey` to a script, which is the cheapest and (maybe) most common case. Then users with other key types can either obtain one by tweaking (or `assume_dangerous` :P).

We probably also should have a method like this, which takes an xonly key and a merkle root, but it should be called something longer, `new_v1_p2tr_full` or something.

**Comment by apoelstra** [comment]
Done reviewing a2e9d2eae63919c9c5988708b932d7adbece0e60

File: src/blockdata/script.rs
Comment (dr-orlovsky) [reply src/blockdata/script.rs]: Implemented

File: src/util/schnorr.rs
Comment (sanket1729) [reply src/util/schnorr.rs:51]: I don't see why we need TweakedPubkeys in PSBT APIs? I think partial sigs should be `XOnlyPublicKey` -> `SchnorrSig`.

Comment (sanket1729) [reply src/util/schnorr.rs:51]: In the most common case, those XOnlyKeys would be tweaked, but I don't see any benefit for adding that type in psbt API. In my view, the _only_ purpose of `TweakedPublicKey` is to guard users using the address API. 

The keys in other places in the codebase might also be tweaked public keys, but we don't gain anything by having those being tweaked public keys (for example psbt API should only operate on XOnlyKeys).  IMO, we should just remove this commit.

File: src/blockdata/script.rs
Comment (sanket1729) [inline src/blockdata/script.rs:345]: In affd45408adc858af9a9cbc03a61ca526a5d41ad
nit: Would prefer the name be `new_v1_p2tr_script_spend`

File: src/blockdata/script.rs
Comment (sanket1729) [inline src/blockdata/script.rs:401]: In 7dc5554dcbbb8f3275fa78db2a5e100fca9e3706

Why add a new API that is being deprecated? This should be removed.

**Review by sanket1729 — COMMENTED** [review]
I think that the previous API for p2tr that took in `UntweakedPublicKey` was better, but not going to stop this PR on it. As long as we are offered the same functionality, I don't care too much about it.

File: src/blockdata/script.rs
Comment (sanket1729) [reply src/blockdata/script.rs]: >  which is the cheapest and (maybe) most common case.

I think the most common case would be p2tr(UntweakedPublicKey). I don't want to go into API design debates :P , but I think this saves the user from importing a trait `TapTweak` and learning about the trait API in the most common case. For descriptors, tr() descriptor has the internal untweaked key as the argument and the uncommon rawtr() descriptor would have the tweaked key.

I would prefer, `new_v1_p2tr(secp, untweaked_key, Option<MerkleRoot>)` and `new_v1_p2tr_raw(tweaked_key)`

Although this also has the benefit that the users know that the tweaking is being done internally in the API and they are expected to use the tweaked keypairs for signing. 

Not going to hold against this because we are offered the same functionality.

File: src/blockdata/script.rs
Comment (dr-orlovsky) [reply src/blockdata/script.rs:401]: Sorry, my bad. Will remove when will be updating the PR according to the results of other discussions here.

**Comment by apoelstra** [comment]
@Kixunil I like this idea but one thing that will complicate the API is that sometimes you need a libsecp context (for tweaking untweaked keys) and sometimes you don't (for no-opping already tweaked keys).

**Comment by dr-orlovsky** [comment]
There is even more complication here: sometimes we need to return tuple with parity flag - and sometimes not. So doing this way will in fact complicate API instead of simplifying: you will have to write `impl IntoTweaked<TweakedAux = (TweakedPubkey, Parity), TweakedKey = TwrakedPubkey>` which is just terrible and still will not let you generalize on function args.

**Comment by apoelstra** [comment]
I think @Kixunil's proposal was simpler: to have a trait `IntoTweaked` which is basically a variant of `Into<TweakedPubkey>`. Then `TweakedPubkey`, `UntweakedPubkey`, `XOnlyPubkey`, `(XOnlyPubkey, MerkleRoot)`, etc., could all implement this.

Then `Script::new_p2tr` could take an arbitrary implementtor of the trait, convert this to a tweaked key, and prefix it with a 1.

Parity would never come into it.

**Comment by dr-orlovsky** [comment]
Then such `IntoTweaked` should not cover `KeyPair`, since there it will be a bad practice to allow providing secret keys for address/script constructors.

**Comment by sanket1729** [comment]
I think we should only strive for the simple possible API in the most common case which I _think_ is using `UnTweakedPublicKey` for almost all wallets. Normally a wallet would,

- derive a BIP32 key
- consider that as `UnTweakedPublicKey` and (if required) compute the tweak. See BIP86 for example.
- Do the tweaking to compute output key
- Encode the output key in an address

I did not check bitcoin core, but I think they probably do something similar. 

I think the already existing API does a good job at combing steps 3 and 4. The user does not have to understand tweaking or import any traits to do the tweaking. The advanced users who do(or do not) do their own tweaking can use the  p2tr_twaeked API. This does also cleanly separates when the secp context is required or not. 
```rust
    /// Create a pay to taproot payload from untweaked key
    pub fn p2tr<C: Verification>(
        secp: &Secp256k1<C>,
        internal_key: UntweakedPublicKey,
        merkle_root: Option<TapBranchHash>,
        );
        
     pub fn p2tr_tweaked(output_key: TweakedPublicKey) -> ;
```
However, if the goal is to allow and force users to do the tweaking so that they remember that private keypairs also need to be tweaked in order to spend it, maybe we want to go the other route.

**Comment by Kixunil** [comment]
True, needing parameters complicates my idea to the point I'm not convinced it's beneficial. One other idea I had was implementing it for tuple `(UnweakedKey, &Secp256k1, Option<MerkleRoot>)` but even though [**not** unprecedented](https://doc.rust-lang.org/std/net/trait.ToSocketAddrs.html#implementors), it's uncommon.

Given that it's impossible to untweak a tweaked pubkey our only options are having a function accepting `TweakedPublicKey` or both. (The other being a convenience function.) I don't care about this much.

**Comment by dr-orlovsky** [comment]
Let me try to summarize results of the discussion above, so we can move forward with finalizing this PR. Feel free to correct me if I understood something wrong.

1. We can't have just a single catch-all method for constructing taproot outputs, taking some sort of generic arguments - since sometimes we need secp contexts, sometimes we do not want to accept key pairs with secret keys etc.
2. Thus, we end up having three main cases:
    2.1. Construct `scriptPubkey` for a given *untweaked* key, as used by wallets for self-transfers and change addresses (which BIP32-derive this untweaked key - or generate it via MuSig) and optional script Merkle root; this includes both descriptor- and non-descriptor based wallets and covers use of `tr` descriptor;
    2.2. Construct `scriptPubkey` for a given *tweaked* key and optional Merkle root as used by wallets when paying to address - but for that we probably do not need a method on Script but instead should use `Address::script_pubkey`; this probably should cover rare use of `rawtr`, which should be converted to address first - and then to `script Pubkey`.
    2.3. Other edge cases, like adding P2C commitments to the internal key. If fact, this is covered if `Script:p2tr` takes *untweaked* key as discussed above.
3. All these cases can be covered by adding the following methods to the `Script` type:
   - `Script::p2tr(secp, untweaked_key, Option<merkle_root>)`
   - `Script::to_p2tr(secp, untweaked_key)`, which is just a convenience method for constructing single-branched script spending paths
   - and we already have `Address::script_pubkey` for the rest (paying to address, using `rawtr`)

**Comment by Kixunil** [comment]
Converting to address is unexpected but I can see why it could be the right thing to do. Is the overhead vs generating the script directly small?

**Comment by apoelstra** [comment]
Okay, how about
* `Script::new_v1_p2tr` which takes an untweaked key (and a secp context)
* `Script::new_v1_p2tr_full` which also takes a merkle root
* `Script::new_v1_p2tr_tweaked` which takes a tweaked key and nothing else

Then I guess `Address` ought to have analogous methods?

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs:51]: We will still need this for `rawtr` processing

**Comment by dr-orlovsky** [comment]
@apoelstra actually, we already kinda have the same methods in `Address` - with the only exception that `new_v1_p2tr` and `new_v1_p2tr_full` are the same method taking optional for merke root. Do we need to separate them as well?

**Comment by apoelstra** [comment]
Oh :) actually if the address methods take an optional merkle root then the `Script` methods should too, for the sake of consistency. Sorry to keep going back and forth like this.

**Comment by dr-orlovsky** [comment]
@apoelstra done

File: src/util/schnorr.rs
Comment (apoelstra) [inline src/util/schnorr.rs:100]: In 5321bd210897265a291540f43e83294149c10af6:

more trailing spaces.

**Review by apoelstra — APPROVED** [review]
utACK 79874ea0036c0bee1e07f9c60e3c9223d15fbfb3

File: src/util/schnorr.rs
Comment (sanket1729) [reply src/util/schnorr.rs:51]: We can parse `XOnlyPubKey`, and use the `assume_dangerous_tweaked` API? I think that is preciously how `rawtr` should work (if it is added in spec). 
I feel strongly about this because we did put so much into Tweaking API, and added type safety with `TweakedPublicKey`. It would be a shame if we can break it this easily.

**Review by sanket1729 — DISMISSED** [review]
ACK 79874ea0036c0bee1e07f9c60e3c9223d15fbfb3. I feel strongly about not having `TweakedPublicKey::from_str` especially when it is really not needed. So far, I don't see any compelling reason to support it apart from it being a bit unergonomic to call (`assume_dangerous`) downstream. I feel that we are adding a foot gun here when there is no need for it. 

But not going to hold this PR hostage because of it.

**Comment by dr-orlovsky** [comment]
I am fine with removing `FromStr`, if other feels that this is more a foot-gun rather than helper to situations like `rawtr`. I rather have no strong opinion on that. 

The only thing I am asking is to do that post-merge in a separate PR. I will open one to discuss that specific matter not to force re-reviwing this one once again.

**Comment by sanket1729** [comment]
> sanket1729 dismissed apoelstra’s stale review via 79874ea 41 seconds ago

I did not do that. I don't know how that happened :(

**Comment by sanket1729** [comment]
@dr-orlovsky, looks there are conflicts here now :( . Can you please separate the first commit into another PR? I will ACK the remaining PR asap

**Comment by dr-orlovsky** [comment]
@sanket1729 done: #772

Will rebase this one w/o that commit after #762 merge

**Review by nlanson — COMMENTED** [review]
Left a small documentation suggestion.

**Comment by sanket1729** [comment]
@dr-orlovsky , #762 merged

**Comment by dr-orlovsky** [comment]
@sanket1729 done

File: src/util/schnorr.rs
Comment (apoelstra) [reply src/util/schnorr.rs:93]: In fad36dc6: 

`keys` should be `key`

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs]: Fixed here and in the other place down. Force-pushed

Comment (apoelstra) [reply src/util/schnorr.rs]: Thanks, looks good, utACK 7405836411130255a4148b31789f95e50c19c3b4

File: src/blockdata/script.rs
Comment (Kixunil) [inline src/blockdata/script.rs:349]: I think this comment here is not needed since tweaking is tracked in the type. Such comment is more appropriate for `assume_tweaked`.

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs]: It seems you forgot to use "Approve" GitHub function

File: src/blockdata/script.rs
Comment (Kixunil) [inline src/blockdata/script.rs:417]: I think `self.0.get(0)?` would be better to pass the responsibility for checking to the caller.

File: src/util/schnorr.rs
Comment (Kixunil) [inline src/util/schnorr.rs:58]: As far as I understand this is an extension trait, so per Rust convention it should be called `KeyExt` or something. (Yes, it's a bit strange that it's an extension of two types actually.)

File: src/util/schnorr.rs
Comment (Kixunil) [inline src/util/schnorr.rs:134]: This clone is confusing, why not use `mut self` or `let mut output_key = self`?

**Review by sanket1729 — APPROVED** [review]
ACK 7405836411130255a4148b31789f95e50c19c3b4

**Review by apoelstra — APPROVED** [review]
ACK 7405836411130255a4148b31789f95e50c19c3b4

File: src/util/schnorr.rs
Comment (Kixunil) [inline src/util/schnorr.rs:140]: I think it'd be cleaner to implement these in terms of public API of `TweakedPublicKey`/`TweakedKeyPair`. Maybe even put them into separate modules to enforce this. Not very important but though.

File: src/util/schnorr.rs
Comment (Kixunil) [inline src/util/schnorr.rs:189]: Should we also `impl From<TweakedKeyPair> for KeyPair`?
Also maybe `into_raw` is a clearer name?

**Review by Kixunil — COMMENTED** [review]
Some style comments otherwise looks good. Not rejecting but would like to see some answer to the comments before I ack.

File: src/util/schnorr.rs
Comment (apoelstra) [reply src/util/schnorr.rs:58]: I think extension traits are only about extending other *traits*.

Comment (Kixunil) [reply src/util/schnorr.rs:58]: Not really, they are very often about extending types. [Even in `std`](https://doc.rust-lang.org/std/os/unix/ffi/trait.OsStrExt.html)

**Comment by dr-orlovsky** [comment]
@Kixunil @apoelstra I propose to merge this PR and address @Kixunil comments right away in a follow-up PR - otherwise I am afraid to have an endless cycle of re-ACKing due to a style/naming changes and API design play.

**Comment by apoelstra** [comment]
Ok by me.

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs:58]: This looks like a single-method extensions, which are usually named after the method name. The second method ("dangerous" one) is dangerous and I think changing the trait name will be more confusing than leaving as is ("KeyExt - extension to what? Which key types? Or extended keys?" And there may be other extenders in the future...).

File: src/blockdata/script.rs
Comment (dr-orlovsky) [reply src/blockdata/script.rs:349]: The commends are different between types, at least should be

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs:140]: Not sure I understood

File: src/util/schnorr.rs
Comment (dr-orlovsky) [reply src/util/schnorr.rs:189]: Will do a From impl in the follow-up PR

File: src/util/schnorr.rs
Comment (Kixunil) [reply src/util/schnorr.rs:58]: Yeah, true neither name is great.

File: src/blockdata/script.rs
Comment (Kixunil) [reply src/blockdata/script.rs:349]: I mean you don't have a "make sure this is dereferencable" comment on `&T` documentation because the type already guarantees it and you have that comment on the "constructor".

File: src/util/schnorr.rs
Comment (Kixunil) [reply src/util/schnorr.rs:140]: I mean calling `TweakedPublicKey::dangerous_assume_tweaked(self)`.

Comment (dr-orlovsky) [reply src/util/schnorr.rs:140]: Still not get it. We can't call `TweakedPublicKey::dangerous_assume_tweaked(self)` since we need key pair, not a public key

**Comment by dr-orlovsky** [comment]
@Kixunil addressed all your comments in #778

**Pull request merged**
